src/stacked_borrows/diagnostics.rs

```diff
@@ -0,0 +1,299 @@
+use rustc_middle::mir::interpret::{AllocId, AllocRange};
+use rustc_span::{Span, SpanData};
+use rustc_target::abi::Size;
+
+use crate::helpers::HexRange;
+use crate::stacked_borrows::{err_sb_ub, AccessKind, GlobalStateInner, Permission};
+use crate::Item;
+use crate::SbTag;
+use crate::Stack;
+
+use rustc_middle::mir::interpret::InterpError;
+
+#[derive(Debug, Default)]
+pub struct AllocHistory {
+    // The time tags can be compressed down to one bit per event, by just storing a Vec<u8>
+    // where each bit is set to indicate if the event was a creation or a retag
+    current_time: usize,
+    creations: smallvec::SmallVec<[Event; 2]>,
+    invalidations: smallvec::SmallVec<[Event; 1]>,
+    protectors: smallvec::SmallVec<[Protection; 1]>,
+}
+
+#[derive(Debug)]
+struct Protection {
+    orig_tag: SbTag,
+    tag: SbTag,
+    span: Span,
+}
+
+#[derive(Debug)]
+struct Event {
+    time: usize,
+    parent: Option<SbTag>,
+    tag: SbTag,
+    range: AllocRange,
+    span: Span,
+}
+
+pub enum TagHistory {
+    Tagged {
+        tag: SbTag,
+        created: (AllocRange, SpanData),
+        invalidated: Option<(AllocRange, SpanData)>,
+        protected: Option<(SbTag, SpanData, SpanData)>,
+    },
+    Untagged {
+        recently_created: Option<(AllocRange, SpanData)>,
+        recently_invalidated: Option<(AllocRange, SpanData)>,
+        matching_created: Option<(AllocRange, SpanData)>,
+        protected: Option<(SbTag, SpanData, SpanData)>,
+    },
+}
+
+pub trait GlobalStateExt {
+    fn add_creation(
+        &mut self,
+        parent: Option<SbTag>,
+        tag: SbTag,
+        alloc: AllocId,
+        range: AllocRange,
+    );
+
+    fn add_invalidation(&mut self, tag: SbTag, alloc: AllocId, range: AllocRange);
+
+    fn add_protector(&mut self, orig_tag: SbTag, tag: SbTag, alloc: AllocId);
```

Comment:
Please call these "history_add_protector" or so; that makes it more clear when reading stacked_borrows.rs that this is all just for diagnostics.

Reply:
I changed these to `log_`, is that clear enough?

Reply:
Yes I like `log`, but then please also rename `get_stack_history` to `get_stack_log` or so?